tower/src/limit/concurrency/layer.rs

```diff
@@ -1,24 +1,67 @@
+use std::sync::Arc;
+
 use super::ConcurrencyLimit;
+use tokio::sync::Semaphore;
 use tower_layer::Layer;
 
 /// Enforces a limit on the concurrent number of requests the underlying
 /// service can handle.
 #[derive(Debug, Clone)]
-pub struct ConcurrencyLimitLayer {
-    max: usize,
+pub struct ConcurrencyLimitLayer<T = usize> {
+    semaphore: T,
 }
 
 impl ConcurrencyLimitLayer {
     /// Create a new concurrency limit layer.
     pub fn new(max: usize) -> Self {
-        ConcurrencyLimitLayer { max }
+        ConcurrencyLimitLayer { semaphore: max }
+    }
+}
+
+impl<T> ConcurrencyLimitLayer<T>
+where
+    T: ToOwnedSemaphore,
+{
+    /// Create a new concurrency limit layer from a type that implements [`ToOwnedSemaphore`]
+    pub fn new_with_semaphore(semaphore: T) -> Self {
+        ConcurrencyLimitLayer { semaphore }
```

Comment:
Not totally convinced it's necessary for this to be totally generic...we could just have a `ConcurrencyLimit::new(usize)` and a `ConcurrencyLimit::new_global(semaphore: Arc<Semaphore>)` or something like that? This would make it clearer to the user that one constructor is used for creating a separate concurrency limit for each service produced by the layer, and the other shares one limit across all such services produced by the layer.

I'm not sure if I can immediately come up with any use-cases for users implementing `ToOwnedSemaphore` for types other than `usize` or `Arc<Semaphore>`...but if there is a reason to, we can always add a more generic constructor later?